### PR TITLE
bpo-35589: Prevent buffer copy in sock_sendall()

### PR DESCRIPTION
**Pull request by asvetlov**
No NEWs is needed since the problem was introduced on master only and never released.

https://bugs.python.org/issue35589

### CODE REVIEW

**Review by eamanu — APPROVED** [review]
LGTM

**Comment by asvetlov** [comment]
There are tests for sending large data already.
An explicit test for memory consumption is not robust.
We can add a mocked test but the test is useless almost: it checks mocks correctness only :)

File: Lib/asyncio/selector_events.py [hunk 1]
```diff
@@ -428,32 +428,34 @@ def _sock_recv_into(self, fut, sock, buf):
...
 
-    def _sock_sendall(self, fut, sock, data):
+    def _sock_sendall(self, fut, sock, data, pos):
         if fut.done():
             # Future cancellation can be scheduled on previous loop iteration
             return
+        start = pos[0]
         try:
-            n = sock.send(data)
+            n = sock.send(data[start:])
```
Comment (1st1) [inline Lib/asyncio/selector_events.py]: You're still copying data here.  Don't you want to wrap it in a memoryview (`memoryview(data)[start:]`)?  Or is it slower that way?

File: Lib/asyncio/selector_events.py [hunk 2]
```diff
@@ -428,32 +428,34 @@ def _sock_recv_into(self, fut, sock, buf):
...
-        else:
-            data = bytearray(memoryview(data)[n:])
 
         fut = self.create_future()
         fd = sock.fileno()
         fut.add_done_callback(
             functools.partial(self._sock_write_done, fd))
-        self.add_writer(fd, self._sock_sendall, fut, sock, data)
+        # use a trick with a list in closure to store a mutable state
+        self.add_writer(fd, self._sock_sendall, fut, sock, data, [n])
```
Comment (1st1) [inline Lib/asyncio/selector_events.py]: But why do we want to have a mutable state here?

File: Lib/asyncio/selector_events.py (hunk 1 above)
Comment (asvetlov) [reply Lib/asyncio/selector_events.py]: `memoryview` is a great suggestion!

File: Lib/asyncio/selector_events.py (hunk 2 above)
Comment (asvetlov) [reply Lib/asyncio/selector_events.py]: I cannot pass just `n` -- need either a mutable object or re-add a writer after every `sock.send()` call until all data is sent.

**Review by methane — APPROVED** [review]
Theorically speaking, memoryview should be closed after using.
But it is ignored often.  Unlike file.close(), not closing memoryview is not a real problem on PyPy.

I want portable version of `os.pwrite(fd, buf, start)`, instead of abusing memoryview though...

**Review by 1st1 — APPROVED** [review]

**Comment by asvetlov** [comment]
Sorry, forgot to merge it 3 months ago

**Pull request merged**